kratos/containers/pointer_vector_set.h

```diff
@@ -647,10 +649,45 @@ class PointerVectorSet final
     template <class InputIterator>
     void insert(InputIterator first, InputIterator last)
     {
-        // first sorts the input iterators and make the input unique.
-        std::sort(first, last, CompareKey());
-        auto new_last = std::unique(first, last, EqualKeyTo());
-        SortedInsert(first, new_last);
+        // We copy always the input range to a temp not to have the input range mutated.
+        std::vector<TPointerType> temp;
+        temp.reserve(std::distance(first, last));
+        for (auto it = first; it != last; ++it) {
+            temp.push_back(GetPointer(it));
+        }
+        std::sort(temp.begin(), temp.end(), CompareKey());
+        auto new_last = std::unique(temp.begin(), temp.end(), EqualKeyTo());
+        SortedInsert(temp.begin(), new_last);
+
+        // KRATOS_ERROR << "HELLO THERE";
+    }
+
+
+    /**
+     * @brief Inserts elements from a r_value container.
+     * @details This method is used to insert all values from an r valued container.
+     *          This mutates the r valued input container to be sorted and unique if it is not the type of PointerVectorSet.
+     *          If it is the type of the PointerVectorSet, then it doesn't mutate the input container.
+     *
+     * @warning The move assumes the ownership of the container, hence the the rContainer may be mutated. if the TContainer is of type std::shared_ptr or intrusive_ptr
+     *          then, there will always be a null pointer at the end past position of the unique sorted list since the std::unique uses the move assignment operator
+     *          within its algorithm. Therefore, the new_last created here will have the correct pointer to the object, and the rContainer will have a nullptr in the
+     *          corresponding place. Therefore, if this method is called once, and if it used with a container type which is not PointerVectorSet, please do not use
+     *          the input rContainer anymore. It will segfault unless the PointerVectorSet is created with raw pointers. [This is because since c++11 std::unique uses the move assignment operator]
+     * @tparam TContainer
+     * @param rContainer
+     */
```

Comment:
The user doesn't have to know all these details. In fact they **must** not know in my opinion. All they should be aware of, is that they moved their container into the function, so they cannot use it anymore.

```suggestion
    /**
     * @brief Inserts items from an @a rvalue container.
     * @details This method takes ownership of the input container, and inserts all its items into @ref PointerVectorSet.
     *
     * @warning The move assumes the ownership of the container, hence the rContainer may be mutated. 
     * @tparam TContainer
     * @param rContainer
     */
```

Reply:
I would say, lets keep this detailed description so that we will find it easier to understand. 99% of the users will never look at the PVS, they will stop at the `ModelPart`.